keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,378 @@
+    RuntimeProfiles []string
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
+
+## Design Details
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be taken for migrating
+to the API fields. The cases to consider are: pod create, pod update, PSP create, PSP update.
+
+All API skew is resolved in the API server. New Kubelets will only use the seccomp values specified
+in the fields, and ignore the annotations.
+
+#### Pod Creation
+
+If no seccomp annotations or fields are specified, no action is necessary.
+
+If _only_ seccomp fields are specified, add the corresponding annotations. This ensures that the
+fields are enforced even if the node version trails the API version. Since [we
+support](https://kubernetes.io/docs/setup/release/version-skew-policy/) up to 2 minor releases of
+version skew between the master and node, this behavior will be removed in version N+2 (where N is
+the release with the upgraded seccomp support).
+
+If _only_ seccomp annotations are specified, copy the values into the corresponding fields. This
+ensures that existing applications continue to enforce seccomp, and prevents the kubelet from
+needing to resolve annotations & fields.
+
+If both seccomp annotations _and_ fields are specified, the values MUST match. This will be enforced
+in API validation.
+
+#### Pod Update
+
+The seccomp fields on a pod are immutable.
+
+The behavior on annotation update is currently ill-defined: the annotation update is allowed, but
+the new value will not be used until the container is restarted. There is no way to tell (from the
+API) what value a container is using.
+
+Therefore, seccomp annotation updates will be ignored. This maintains backwards API compatibility
+(no tightening validation), and makes a small stabilizing change to behavior (new Kubelets will
+ignore the update).
+
+#### PodSecurityPolicy Creation
+
+_Similar policy to [Pod Creation](#pod-creation)_
+
+If no seccomp annotations or fields are specified, no action is necessary.
+
+If _only_ seccomp fields are specified, add the corresponding annotations. This ensures that the
+fields are enforced even in skewed high-availability environments. Since [we
+support](https://kubernetes.io/docs/setup/release/version-skew-policy/) up to 1 minor release of
+version skew between the HA masters, this behavior will be removed in version N+1 (where N is the
+release with the upgraded seccomp support).
+
+If _only_ seccomp annotations are specified, copy the values into the corresponding fields. This
+smooths the migration for a future release that removes annotation support, and frees tools from
+needing to check annotations.
+
+If both seccomp annotations _and_ fields are specified, the values MUST match. This will be enforced
+in API validation.
+
+#### PodSecurityPolicy Update
```

Comment:
need to consider clients that update and drop the new field because they are not aware of it. each field has three options: unchanged, changed, removed. we need to consider all 9 cases of these interactions

Reply:
as an alternative, consider keeping the fields and the annotations parallel, and not defaulting one from the other (two-way defaulting on mutable fields is nasty)